network.go

```diff
@@ -1032,142 +1032,61 @@ func createNetworkEndpoints(numOfEndpoints int) (endpoints []Endpoint, err error
 	return endpoints, nil
 }
 
-func getIfaceByNameFromNetNs(networkNSPath string, ifaceName string) (NetIfaceAddrs, error) {
-	var netIface NetIfaceAddrs
-
-	err := doNetNS(networkNSPath, func(_ ns.NetNS) error {
-		iface, err := net.InterfaceByName(ifaceName)
-		if err != nil {
-			return err
-		}
-		if iface == nil {
-			return fmt.Errorf("Interface is nil")
-		}
-
-		addrs, err := iface.Addrs()
-		if err != nil {
-			return err
-		}
-
-		var ipNetList []net.IPNet
-		for _, addr := range addrs {
-			ipNet, ok := addr.(*net.IPNet)
-			if !ok {
-				continue
-			}
-
-			ipNetList = append(ipNetList, *ipNet)
-		}
-
-		netIface = NetIfaceAddrs{
-			Iface: *iface,
-			Addrs: ipNetList,
-		}
-
-		return nil
-	})
+func networkInfoFromLink(handle *netlink.Handle, link netlink.Link) (NetworkInfo, error) {
+	addrs, err := handle.AddrList(link, netlink.FAMILY_ALL)
 	if err != nil {
-		return NetIfaceAddrs{}, err
+		return NetworkInfo{}, err
 	}
 
-	return netIface, nil
-}
-
-func getIfacesFromNetNsFilter(networkNSPath string, ipFilter bool) ([]NetIfaceAddrs, error) {
-	var netIfaces []NetIfaceAddrs
-
-	if networkNSPath == "" {
-		return []NetIfaceAddrs{}, fmt.Errorf("Network namespace path cannot be empty")
+	routes, err := handle.RouteList(link, netlink.FAMILY_ALL)
+	if err != nil {
+		return NetworkInfo{}, err
 	}
 
-	err := doNetNS(networkNSPath, func(_ ns.NetNS) error {
-		ifaces, err := net.Interfaces()
-		if err != nil {
-			return err
-		}
+	attrs := link.Attrs()
 
-		for _, iface := range ifaces {
-			addrs, err := iface.Addrs()
-			if err != nil {
-				return err
-			}
-
-			if ipFilter {
-				// Ignore unconfigured network interfaces
-				// These are either base tunnel devices
-				// that are not namespaced like
-				// gre0, gretap0, sit0, ipip0, tunl0
-				// or incorrectly setup interfaces
-				if (addrs == nil) || (len(addrs) == 0) {
-					continue
-				}
-			}
-
-			var ipNetList []net.IPNet
-			for _, addr := range addrs {
-				ipNet, ok := addr.(*net.IPNet)
-				if !ok {
-					continue
-				}
-
-				ipNetList = append(ipNetList, *ipNet)
-			}
-
-			netIface := NetIfaceAddrs{
-				Iface: iface,
-				Addrs: ipNetList,
-			}
-
-			netIfaces = append(netIfaces, netIface)
-		}
+	return NetworkInfo{
+		Iface:  *attrs,
+		Addrs:  addrs,
+		Routes: routes,
+	}, nil
+}
 
-		return nil
-	})
+func networkInfoFromIfaceName(handle *netlink.Handle, ifaceName string) (NetworkInfo, error) {
```

Comment:
An observation. If we have the link index cached we should use the link index. The reason is that the LinkByName is highly inefficient and take much longer to complete. So we should see if we can use the index as far as possible when it is available, so that network setup does not contribute to start latency.

Reply:
This function is removed in another commit, no need to worry about it, but thanks for the heads up :)